### PR TITLE
H294: Student-t (df=3) weight-noise at EP15+anti-K3+6-res — heaviest finite-var tail

### PR DESCRIPTION
**Pull request by morganmcg1**
# H294: Student-t (df=3) noise at EP15+anti-K3+6-res — does heaviest finite-variance tail improve channel diversity?

## Hypothesis

**Test Student-t (df=3) weight noise at matched RMS at EP15+anti-K3+6-res — does heaviest finite-variance noise distribution add complementary gradient direction diversity that Laplace and Gaussian miss?**

Your H287 (EP14, just closed) banked Finding III-EP-axis-fully-exhausted. Combined with edward's H284 (Finding GGG-Sobol×anti non-stacking) and nezuko's H280 (Finding HHH-EP13-no-rescue), the EP and Sobol axes are now fully closed. We need orthogonal axes.

**Noise distribution family is untouched.** Tail progression:
- Gaussian: P(|δ| > t) ~ exp(-t²/2σ²) — very thin tail (current SOTA)
- Laplace: P(|δ| > t) ~ exp(-t/b) — exponential tail (nezuko H293 parallel)
- **Student-t (df=3): P(|δ| > t) ~ t^-(df+1)** — polynomial tail, much heavier than exponential
- Matched RMS: Student-t df=3 scales by σ·√((df-2)/df) = σ/√3 so RMS_δ = σ

With df=3, Student-t has finite variance but infinite higher moments — meaning **rare but very large excursions in some weight dimensions per sample**. The anti-thetic ±δ structure preserves the linear Taylor cancellation regardless of distribution, but the higher-order residuals can be qualitatively different under heavy tail.

**The hypothesis**: If the channel-asymmetric pattern (Finding XX: SP/VP-favorable σ=3e-4 vs WSS-favorable σ=5e-4) comes from rare-but-important weight perturbations that probe specific directions in the loss landscape, **a heavier-tail sampler may hit those directions more often**, improving channel-asymmetric coverage even at single σ.

Three concrete predictions:
1. **Student-t wins**: val < 5.9243 AND test < 5.7690 → heavy-tail captures rare critical directions → SOTA + bank Finding "Student-t-heavy-tail-superior"
2. **Student-t ≈ Gaussian**: tail behavior doesn't matter at fixed σ → bank "noise-tail-axis-null"; confirms Finding XX σ-variance is the only source of channel asymmetry
3. **Student-t worse**: heavy tail introduces unstable forward-pass amplifications → confirms thin-tail Gaussian is structurally optimal

**Predicted val_abupt: 5.920-5.935**.

## Why this is the right next step for you

- You just closed H287 (EP14, val 5.9326/test 5.7764); EP-axis exploration is done
- The Sobol-anti axis is closed at K=3 EP15 (Finding GGG from edward H284)
- The natural next axis is **noise distribution family** — completely untested
- You ran the same EP15+anti-K3+6-res recipe in H287 (just with EP14 checkpoint), so you know the infrastructure
- Single code change: add Student-t sampler to the noise function

## Why complementary to nezuko H293

| Distribution | Tail | Sampler | Predicted gain mechanism |
|---|---|---|---|
| Gaussian (H275 SOTA) | Thin exp(-t²) | torch.randn | reference |
| Laplace (H293 nezuko) | Medium exp(-t) | inverse CDF | moderate heavy tail |
| **Student-t df=3 (you)** | **Heavy t^-4** | **z/sqrt(χ²/df)** | **rare-but-large excursions** |

H293 + H294 together fully characterize the noise-tail axis. If both Laplace and Student-t lose to Gaussian, noise distribution is fully closed. If Laplace ≈ Gaussian but Student-t wins, the result tells us the gain comes from rare extreme events, not moderate excursions.

### Step 1: Branch off latest tay

```bash
git fetch origin tay
git checkout -b askeladd/h294-student-t-noise-ep15-anti-K3 origin/tay
```

### Step 2: Patch eval_tta_h252.py to support `--noise-distribution`

Coordinate with nezuko's H293 — if their PR is already merged to tay, you'll just use their flag. If not, add the same flag yourself:

```python
parser.add_argument("--noise-distribution", type=str, default="gaussian",
                    choices=["gaussian", "laplace", "student_t"])
parser.add_argument("--student-t-df", type=float, default=3.0)

def sample_weight_noise(param_shape, sigma, distribution="gaussian", df=3.0, device="cuda", generator=None):
    if distribution == "gaussian":
        return torch.randn(param_shape, device=device, generator=generator) * sigma
    elif distribution == "laplace":
        b = sigma / math.sqrt(2)
        u = torch.rand(param_shape, device=device, generator=generator) - 0.5
        return -b * torch.sign(u) * torch.log(1 - 2 * u.abs())
    elif distribution == "student_t":
        assert df > 2, f"Student-t df must be > 2 for finite variance, got {df}"
        scale = sigma * math.sqrt((df - 2) / df)
        z = torch.randn(param_shape, device=device, generator=generator)
        chi2 = torch.distributions.Chi2(df).sample(param_shape).to(device)
        return scale * z / torch.sqrt(chi2 / df)
```

**Verify RMS calibration**: print `noise.pow(2).mean().sqrt()` for Student-t df=3 at σ=5e-4 over a large tensor — should return ~5e-4.

### Step 3: Run primary command

```bash
H185_EP15_CKPT="outputs/drivaerml/run-0gjfv45i/checkpoint_ep15.pt"

torchrun --standalone --nproc-per-node=8 eval_tta_h252.py \
  --checkpoint $H185_EP15_CKPT \
  --resolutions "32768,49152,65536,81920,98304,131072" \
  --eval-modes "weight_noise_mirror_res_avg" \
  --weight-noise-sigma 5e-4 \
  --weight-noise-passes 3 \
  --antithetic-noise \
  --noise-distribution student_t \
  --student-t-df 3.0 \
  --batch-size 2 --num-workers 4 \
  --agent askeladd \
  --wandb-group "h294-askeladd-ep15-anti-K3-student-t" \
  --wandb-name "askeladd/h294-ep15-anti-K3-student-t-df3-stack"
```

K=3 anti-thetic = 6 forward passes. Same compute as H275 (~230 min on DDP×8).

### Step 4: Defensive NCCL timeout patch

Same `timedelta(minutes=60)` in `init_distributed`. You have this from H287.

### Step 5: Post terminal SENPAI-RESULT

```
SENPAI-RESULT: {"terminal":true,"status":"complete","pending_arms":false,"wandb_run_ids":["<run-id>"],"primary_metric":{"name":"val_abupt_ep15_anti_K3_student_t_stack","value":<val>},"test_metric":{"name":"test_abupt_ep15_anti_K3_student_t_stack","value":<test>}}
```

Include the channel breakdown table:

| Recipe | val_abupt | test_abupt | test_WSS | test_SP | test_VP |
|---|---:|---:|---:|---:|---:|
| **H294 Student-t df=3 (you)** | **?** | **?** | **?** | **?** | **?** |
| H275 EP15+anti-K=3+σ=5e-4 Gaussian (SOTA) | 5.9243 | 5.7690 | 6.6743 | 3.6427 | 3.3788 |
| H287 EP14+anti-K=3+σ=5e-4 Gaussian (your prior) | 5.9326 | 5.7764 | 6.68 | 3.65 | 3.39 |

Plus 2-3 sentences:
- Did Student-t improve any channel?
- Did the heavy tail introduce instability or improve channel-asymmetric coverage?
- What's the SP-floor delta? (target: ≤ 3.577)

**If val < 5.9243 AND test < 5.7690 → immediate SOTA merge** (especially if SP/VP improve)

## Baseline metrics

| Reference | val_abupt | test_abupt | test_WSS | test_VP | test_SP | W&B |
|---|---:|---:|---:|---:|---:|---|
| **H275 EP15+anti-K3+σ=5e-4 Gaussian ← CURRENT SOTA** | **5.9243** | **5.7690** | **6.6743** | **3.3788** | **3.6427** | 0b4t2bz2 |
| H277 EP15+anti-K3+σ=3e-4 Gaussian | 5.9255 | 5.7705 | 6.6755 | 3.3756 | 3.6406 | mvnwbq3z |
| H287 EP14+anti-K3+σ=5e-4 Gaussian (your prior) | 5.9326 | 5.7764 | — | — | — | x2q6nkqe |
| **Transolver-3 target (Morgan #1056)** | — | — | **< 5.850%** | **≤ 3.421%** | **≤ 3.577%** | — |

**Merge gate**: val_abupt < **5.9243** AND test_abupt < **5.7690**

## Critical constraints

- **DDP 8 GPUs** (`--nproc-per-node=8`) for every eval run
- **Read-only**: data/loader.py, data/preload.py, data/split_manifest.json, model code
- **No model capacity changes** — eval-time TTA only
- **z-axis tau_z LOCKED at 1.67**
- **SENPAI_TIMEOUT_MINUTES=360** (~230 min predicted, same as H275)

## Implementation notes

- **RMS calibration is critical**: verify Student-t at σ=5e-4 RMS via `noise.pow(2).mean().sqrt()` on a 1M-element tensor before launching primary.
- **Stability check**: Student-t with df=3 has finite variance but infinite kurtosis. If you see NaN/Inf in any forward pass during the first few cases, lower σ to 3e-4 and rerun (rename to "h294-student-t-df3-sigma3e-4"). Do not silently truncate.
- **Anti-thetic preservation**: same delta tensor used for +/- pair, just sign-flipped. Sampler called once per pair, not per sign.
- **Seed determinism**: use `torch.Generator(device=cuda).manual_seed(seed_k)` per pair.
- **Per-distribution config logging**: log `noise_distribution=student_t` and `student_t_df=3.0` in W&B config.

## Suggested follow-ups (post-result)

1. If Student-t df=3 wins → assign Student-t df=2.5 (even heavier) to push tail to near-Cauchy
2. If Student-t df=3 ties → noise-tail axis closed; pivot to per-layer noise or noise-anisotropy (per-tensor σ scaling)
3. If Student-t df=3 degrades → confirms Gaussian is structurally optimal; closes noise-family axis fully
4. The H293+H294 pair completes the noise-family characterization for the paper

### CODE REVIEW

**Comment by morganmcg1** [comment]
STUDENT askeladd: RMS calibration check (PR step 2 verification)

Ran `_sample_noise(student_t, df=3.0, sigma=5e-4)` over 5M-element tensor:

| Distribution | Empirical RMS | Target | Shortfall |
|---|---:|---:|---:|
| Student-t df=3 | 4.964e-04 | 5e-04 | 0.72% |
| Gaussian | 5.000e-04 | 5e-04 | 0.00% |

Shortfall is from clamp at ±50 (max |t|/σ = 50.0 hit at fraction 3.6e-6, exceedance rate matches df=3 theory: $P(|T|>50) \approx 0.5 \cdot 50^{-3} \cdot \Gamma(2)/\Gamma(1.5)/\sqrt{3\pi} \approx 3.6e-6$). Calibration confirmed valid for production run.

Primary run launched at 08:43Z (wandb `jb36zg3o`). Will report results when complete.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #1482 because ADVISOR: Closing H294 — Student-t noise family fails val gate.

**Finding QQQ: Student-t (df=3) noise family is functionally equivalent to Gaussian at RMS σ=5e-4 — noise distribution family axis is now definitively closed**

W&B run jb36zg3o val_abupt = **5.9248%** — fails gate by **1.3 bp** (gate: < 5.9235%).

Student-t with df=3 at RMS σ≈4.964e-4 produces essentially the same result as Gaussian noise (Finding MMM, H293 nezuko). The heavier tails of Student-t provide no benefit for anti-thetic TTA noise sampling. This was the last major noise distribution family to test.

**Noise family axis summary:**
- Gaussian (baseline SOTA): val=5.9235% ✓ (PASSES gate)
- Laplace (Finding MMM, H293): val failed gate — null result
- Student-t df=3 (Finding QQQ, H294): val=5.9248% — fails gate by 1.3 bp

**Conclusion:** The noise distribution family does not matter in the anti-thetic TTA framework at RMS σ=5e-4. Gaussian remains optimal (or more precisely, the framework is insensitive to tail behavior). Focus should shift to other axes: noise magnitude (K), resolution configuration, or per-layer stratification.

Current SOTA for reference: H285 EP15+Anti-thetic-K=4+6-res×mirror, val=5.9235%, test=5.7683%..
